addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/adapters/Record.java

```diff
@@ -0,0 +1,139 @@
+                log.trace("Rule {} does not exist", ruleRef);
+                continue;
+            }
+            DocumentModel ruleDoc = session.getDocument(ruleRef);
+            RetentionRule rule = ruleDoc.getAdapter(RetentionRule.class);
+            if (!rule.isEnabled()) {
+                log.debug("Rule {} id disabled", ruleDoc::getPathAsString);
+                continue;
+            }
+            return rule;
+        }
+        log.debug("No active rules found for {}", document::getPathAsString);
+        return null;
+
+    }
+
+    protected List<String> getRuleIds() {
+        Serializable propertyValue = document.getPropertyValue(RetentionConstants.RECORD_RULE_IDS_PROP);
+        if (propertyValue == null) {
+            return Collections.emptyList();
+        }
+        return Arrays.asList((String[]) propertyValue);
+    }
+
+    public Calendar getSavedRetainUntil() {
+        Serializable savedRetainUntil = document.getPropertyValue(RetentionConstants.RETAIN_UNTIL_PROP);
+        if (savedRetainUntil != null) {
+            return (Calendar) savedRetainUntil;
+        }
+        return null;
+    }
+
+    public boolean isRetainUntilInderterminate() {
+        if (!getDocument().isUnderRetentionOrLegalHold()) {
+            return false;
+        }
+        Calendar retainUntil = getDocument().getRetainUntil();
+        return retainUntil != null
+                ? CoreSession.RETAIN_UNTIL_INDETERMINATE.getTimeInMillis() == retainUntil.getTimeInMillis()
+                : false;
+    }
+
+    public boolean isRetentionExpired() {
+        if (!getDocument().isUnderRetentionOrLegalHold()) {
+            return true;
+        }
+        Calendar retainUntil;
+        return (retainUntil = getDocument().getRetainUntil()) == null || !Calendar.getInstance().before(retainUntil);
+    }
+
+    protected void save(CoreSession session) {
+        document.putContextData(DublinCoreListener.DISABLE_DUBLINCORE_LISTENER, true);
+        document.putContextData(NotificationConstants.DISABLE_NOTIFICATION_SERVICE, true);
+        document.putContextData(NXAuditEventsService.DISABLE_AUDIT_LOGGER, true);
+        document.putContextData(VersioningService.DISABLE_AUTO_CHECKOUT, true);
+        document.putContextData(RetentionConstants.RETENTION_CHECKER_LISTENER_IGNORE, true);
+        session.saveDocument(document);
+        document.putContextData(DublinCoreListener.DISABLE_DUBLINCORE_LISTENER, null);
+        document.putContextData(NotificationConstants.DISABLE_NOTIFICATION_SERVICE, null);
+        document.putContextData(NXAuditEventsService.DISABLE_AUDIT_LOGGER, null);
+        document.putContextData(VersioningService.DISABLE_AUTO_CHECKOUT, null);
+        document.getContextData().remove(RetentionConstants.RETENTION_CHECKER_LISTENER_IGNORE);
+    }
+
+    public void saveRetainUntil(Calendar retainUntil) {
+        document.setPropertyValue(RetentionConstants.RETAIN_UNTIL_PROP, retainUntil);
```

Comment:
should use `CoreSession#setRetainUtil` which checks the permissions before setting the value, and the method can  be replaced by `document.getCoreSession().setRetainUntil(getDocument().getRef(), retainUntil, null);` but we should ensure that the `document.getCorSession()` is not `null`

Reply:
But this is an adapter for the current document, and the `RETAIN_UNTIL_PROP` here is not the low-level one you're thinking about.